### PR TITLE
chore(rpc): impl `EthApiServer` helper traits on type like `EthApi`

### PR DESCRIPTION
**Pull request by emhane**
- Moves block building into default impl of `LoadPendingBlock`, so it can be reused for any network stack.
- Extracts methods for assembling receipt and calculating receipts root from block building, so can be overridden for Optimism.
- Introduces macro rules for implementing `EthApiServer` on types with same data layout as `EthApi`. This makes it quick to implement custom `EthApiServer` behaviour for other networks, that are identical in data layout to `EthApi`, like `OptimismApi`.

### CODE REVIEW

File: crates/rpc/rpc/src/eth/api/traits/pending_block.rs [hunk 1]
```diff
@@ -148,4 +166,263 @@ pub trait LoadPendingBlock {
...
+    ) -> Receipt {
+        Receipt {
+            tx_type: tx.tx_type(),
+            success: result.is_success(),
+            cumulative_gas_used,
+            logs: result.into_logs().into_iter().map(Into::into).collect(),
+            #[cfg(feature = "optimism")]
+            deposit_nonce: None,
+            #[cfg(feature = "optimism")]
+            deposit_receipt_version: None,
```
Comment (mattsse) [inline crates/rpc/rpc/src/eth/api/traits/pending_block.rs:185]: we need to find a way to get rid of these features

**Review by mattsse — APPROVED** [review]
the helper macros make sense, to quickly implement the sub trait

File: crates/rpc/rpc/src/eth/api/traits/pending_block.rs (hunk 1 above)
Comment (emhane) [reply crates/rpc/rpc/src/eth/api/traits/pending_block.rs:185]: yeah...I know...

Comment (emhane) [reply crates/rpc/rpc/src/eth/api/traits/pending_block.rs:185]: they should be able to remove when optimism feature is entirely removed from the crate, as long as we have some dep for which we pull in optimism feature, likelihood that that crate pulls in `reth-primitives` with optimism feature, is very high - and we're screwed

**Pull request merged**